Sources/SocksCore/TCPSocket.swift

```diff
@@ -39,22 +39,29 @@ public protocol TCPReadableSocket: TCPSocket { }
 extension TCPReadableSocket {
 
     public func recv(maxBytes: Int = BufferCapacity) throws -> [UInt8] {
+    	var out = Array<UInt8>()
         let data = Bytes(capacity: maxBytes)
         let flags: Int32 = 0 //FIXME: allow setting flags with a Swift enum
         let receivedBytes = socket_recv(self.descriptor, data.rawBytes, data.capacity, flags)
-        guard receivedBytes > -1 else { throw SocksError(.readFailed) }
-        
-        guard receivedBytes > 0 else {
+        guard receivedBytes > -1 else {
+	    if errno == ECONNRESET { // closed by peer, need to close this side too!
+	       _ = try? self.close()
+	       return [] 
+	    } else {
+	      	throw SocksError(.readFailed)
+	    }
+	}
+      
+        if receivedBytes == 0  {
```

Comment:
I don't see a difference in the code here and I prefer the guard since it forces us to leave scope. If there's something I'm missing here, let me know and we can discuss, maybe add a comment.